### PR TITLE
Box-Muller: resample zeros.

### LINKED ISSUE #1464: Overflow in `randn` using CUDA.jl's native RNG
**Describe the bug**

The native CUDA.jl rng regularly produces Infs when sampling standard normal Float32s. This is not expected: for practical purposes, the standard normal distribution has zero measure near `floatmax(Float32)`, and neither CURAND's rng nor the default CPU rng have similar behavior.

**To reproduce**

The Minimal Working Example (MWE) for this bug:

```julia
using CUDA
using Random

println(maximum(randn(CUDA.default_rng(), Float32, 2^26)))         # Likely Inf
println(maximum(randn(CUDA.CURAND.default_rng(), Float32, 2^26)))  # Likely in (5.3, 6.1)
println(maximum(randn(Random.default_rng(), Float32, 2^26)))       # Likely in (5.3, 6.1)
```

<details><summary>Manifest.toml</summary>
<p>

```
CUDA v3.8.5
GPUArrays v8.3.2
GPUCompiler v0.13.14
LLVM v4.9.1
```

</p>
</details>

**Expected behavior**

No overflow, like with `CUDA.CURAND.default_rng()` and `Random.default_rng()`.

**Version info**

Details on Julia:

```
Julia Version 1.7.2
Commit bf53498635 (2022-02-06 15:21 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Platinum 8280 CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, cascadelake)
```

Details on CUDA:

```
CUDA toolkit 11.6, artifact installation
NVIDIA driver 465.19.1, for CUDA 11.3
CUDA driver 11.3

Libraries: 
- CUBLAS: 11.8.1
- CURAND: 10.2.9
- CUFFT: 10.7.0
- CUSOLVER: 11.3.2
- CUSPARSE: 11.7.1
- CUPTI: 16.0.0
- NVML: 11.0.0+465.19.1
- CUDNN: 8.30.2 (for CUDA 11.5.0)
- CUTENSOR: 1.4.0 (for CUDA 11.5.0)

Toolchain:
- Julia: 1.7.2
- LLVM: 12.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device capability support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

2 devices:
  0: NVIDIA A100-PCIE-40GB (sm_80, 35.178 GiB / 39.586 GiB available)
  1: NVIDIA A100-PCIE-40GB (sm_80, 26.063 GiB / 39.586 GiB available)
```

### PR DESCRIPTION
**Pull request by maleadt**
Fixes https://github.com/JuliaGPU/CUDA.jl/issues/1464

@danielwe: Given your explanation in https://github.com/JuliaGPU/CUDA.jl/issues/1464#issuecomment-1090893798, I just check for `zero` instead of `eps`. It also helps for complex numbers, where we don't have an `eps` defined.

### CODE REVIEW

**Comment by danielwe** [comment]
Thanks for getting around to a fix for this. This solution works even if it does discard a little bit of entropy. A few comments, but please note that I'm no RNG expert, even if I have a fairly decent understanding:

* Regarding complex: I just noticed that this implementation is buggy and does not return standard normal complex numbers. The procedure to get a standard normal complex is to sample `U1` and `U2` of type `real(T)`, do Box-Muller, and return `(Z1 + im * Z2) / sqrt(2)` (in practice you can skip the multiplication by `sqrt(2)` in the Box-Muller transform and hence avoid the division by `sqrt(2)` in the final step).
* Hence, the Box-Muller transform itself should only work with real numbers, so there's no concern about not having `eps`.
* Thoughts on the possible solutions to the Inf issue and which is preferable:
  * Rejecting `U1 == 0` does discard some entropy (i.e., reduces the sample space), but is simple and robust and works for every possible variant of `rand`. As for thread synchronization, with 32-thread warps a `zero(Float32)` will only appear in about every 2^18 or 2^19 warp, so even if it's not ideal it shouldn't hurt performance much in practice.
  * Always adding `eps(T) / 2` retains all the entropy as long as you're sampling uniformly in `[0, 1)`, so is probably slightly preferable from a theoretical standpoint in the current case. However, it couples the transform more tightly to the implementation of `rand`, since you'd rather avoid it if `rand` returns samples in `(0, 1]` like the CURAND rng does (in which case it would introduce excess bias), or if it samples on a finer grid closer to zero like the CURAND rng also does (in which case it probably discards more entropy than rejection sampling).
  * Note that the proposal in https://github.com/JuliaLang/julia/issues/44887 does not break the `eps(T) / 2` approach since `rand` will still be sampling on a uniform grid in `[0, 1)`.

So to sum up:

* The implementation needs fixing for complex numbers
* The choice between rejection sampling and adding `eps(T) / 2` is a choice between robustness (never think about this issue again) and retaining all the entropy (but having to change the transform if `rand` no longer uses a uniform grid in `[0, 1)`).

**Comment by maleadt** [comment]
Thanks for the detailed explanation! Let's put both fixes in https://github.com/JuliaGPU/CUDA.jl/pull/1518.
